File: src/lib/live-location-utils.ts
import type { UserLiveLocationShare, ChatLocationShare } from '@/lib/types';

/** Живая трансляция: срок из `liveSession.expiresAt` истёк — UI показывает текст вместо карты ([`MessageLocationCard`](../src/components/chat/parts/MessageLocationCard.tsx)). */
export function isChatLiveLocationShareExpired(share: ChatLocationShare, nowMs: number = Date.now()): boolean {
  if (!share.liveSession) return false;
  if (!share.liveSession.expiresAt) return false;
  return new Date(share.liveSession.expiresAt).getTime() <= nowMs;
}

const LIVE_MESSAGE_SESSION_SLOP_MS = 15_000;

/**
 * Сообщение с `liveSession` ещё должно показывать превью карты (таймер «живой» локации).
 * Учитывает: дедлайн `expiresAt`, активную запись `users/{sender}.liveLocationShare` и совпадение
 * сессии с сообщением (`startedAt`), чтобы после «Остановить» и без правки документа сообщения
 * пузырь перешёл в тот же текстовый вид, что и при истечении таймера.
 */
export function isChatLiveLocationMessageStillStreaming(
  share: ChatLocationShare,
  messageCreatedAt: string,
  senderLiveShare: UserLiveLocationShare | null | undefined,
  senderProfileResolved: boolean,
  nowMs: number = Date.now(),
): boolean {
  if (!share.liveSession) return false;

  const expiresAtIso = share.liveSession.expiresAt;
  if (expiresAtIso && new Date(expiresAtIso).getTime() <= nowMs) {
    return false;
  }

  if (!senderProfileResolved) {
    return true;
  }

  if (!senderLiveShare || !isLiveShareVisible(senderLiveShare, nowMs)) {
    return false;
  }

  const msgMs = new Date(messageCreatedAt).getTime();
  const startedMs = new Date(senderLiveShare.startedAt).getTime();
  if (Number.isFinite(msgMs) && Number.isFinite(startedMs) && msgMs + LIVE_MESSAGE_SESSION_SLOP_MS < startedMs) {
    return false;
  }

  return true;
}

export function isLiveShareExpired(share: UserLiveLocationShare, nowMs: number = Date.now()): boolean {
  if (!share.expiresAt) return false;
  return new Date(share.expiresAt).getTime() <= nowMs;
}

export function isLiveShareVisible(share: UserLiveLocationShare | null | undefined, nowMs?: number): boolean {
  if (!share?.active) return false;
  return !isLiveShareExpired(share, nowMs);
}

File: mobile/app/lib/features/chat/data/live_location_utils.dart
import 'package:lighchat_models/lighchat_models.dart';

const int _liveMessageSessionSlopMs = 15000;

bool isLiveShareExpired(UserLiveLocationShare share, [int nowMs = 0]) {
  final now = nowMs > 0 ? nowMs : DateTime.now().millisecondsSinceEpoch;
  final exp = share.expiresAt;
  if (exp == null || exp.isEmpty) return false;
  final t = DateTime.tryParse(exp)?.millisecondsSinceEpoch;
  if (t == null) return false;
  return t <= now;
}

bool isLiveShareVisible(UserLiveLocationShare? share, [int nowMs = 0]) {
  if (share == null || !share.active) return false;
  return !isLiveShareExpired(share, nowMs);
}

/// Сообщение с `liveSession` ещё показывает превью карты (паритет `live-location-utils.ts`).
bool isChatLiveLocationMessageStillStreaming(
  ChatLocationShare share,
  DateTime messageCreatedAt,
  UserLiveLocationShare? senderLiveShare,
  bool senderProfileResolved, [
  int nowMs = 0,
]) {
  final now = nowMs > 0 ? nowMs : DateTime.now().millisecondsSinceEpoch;
  if (share.liveSession == null) return false;

  final expiresAtIso = share.liveSession!.expiresAt;
  if (expiresAtIso != null && expiresAtIso.isNotEmpty) {
    final exp = DateTime.tryParse(expiresAtIso)?.millisecondsSinceEpoch;
    if (exp != null && exp <= now) return false;
  }

  if (!senderProfileResolved) return true;

  if (!isLiveShareVisible(senderLiveShare, now)) return false;

  final msgMs = messageCreatedAt.millisecondsSinceEpoch;
  final startedMs = DateTime.tryParse(senderLiveShare!.startedAt)?.millisecondsSinceEpoch;
  if (startedMs != null &&
      msgMs + _liveMessageSessionSlopMs < startedMs) {
    return false;
  }

  return true;
}
